File: go/internal/validation/validate_test.go
package validation_test

import (
	"testing"

	"github.com/microsoft/beachball/internal/options"
	"github.com/microsoft/beachball/internal/testutil"
	"github.com/microsoft/beachball/internal/types"
	"github.com/microsoft/beachball/internal/validation"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// get default options for this file (fetch disabled)
func getDefaultOptions() types.BeachballOptions {
	defaultOptions := types.DefaultOptions()
	defaultOptions.Branch = testutil.DefaultRemoteBranch
	defaultOptions.Fetch = false

	return defaultOptions
}

// TS: "succeeds with no changes"
func TestSucceedsWithNoChanges(t *testing.T) {
	factory := testutil.NewRepositoryFactory(t, "monorepo")
	repo := factory.CloneRepository()
	repo.Checkout("-b", "test", testutil.DefaultBranch)

	buf := testutil.CaptureLogging(t)
	repoOpts := getDefaultOptions()

	parsed := options.GetParsedOptionsForTest(repo.RootPath(), types.CliOptions{}, repoOpts)
	result, err := validation.Validate(parsed, validation.ValidateOptions{
		CheckChangeNeeded: true,
	})
	require.NoError(t, err)
	assert.False(t, result.IsChangeNeeded)
	assert.Contains(t, buf.String(), "Validating options and change files...")
}

// TS: "exits with error by default if change files are needed"
func TestExitsWithErrorIfChangeFilesNeeded(t *testing.T) {
	factory := testutil.NewRepositoryFactory(t, "monorepo")
	repo := factory.CloneRepository()
	repo.Checkout("-b", "test", testutil.DefaultBranch)
	repo.CommitChange("packages/foo/test.js")

	buf := testutil.CaptureLogging(t)
	repoOpts := getDefaultOptions()

	parsed := options.GetParsedOptionsForTest(repo.RootPath(), types.CliOptions{}, repoOpts)
	_, err := validation.Validate(parsed, validation.ValidateOptions{
		CheckChangeNeeded: true,
	})
	assert.Error(t, err)
	assert.Contains(t, buf.String(), "ERROR: Change files are needed!")
	assert.Contains(t, buf.String(), "Found changes in the following packages")
}

// TS: "returns and does not log an error if change files are needed and allowMissingChangeFiles is true"
func TestReturnsWithoutErrorIfAllowMissingChangeFiles(t *testing.T) {
	factory := testutil.NewRepositoryFactory(t, "monorepo")
	repo := factory.CloneRepository()
	repo.Checkout("-b", "test", testutil.DefaultBranch)
	repo.CommitChange("packages/foo/test.js")

	buf := testutil.CaptureLogging(t)
	repoOpts := getDefaultOptions()

	parsed := options.GetParsedOptionsForTest(repo.RootPath(), types.CliOptions{}, repoOpts)
	result, err := validation.Validate(parsed, validation.ValidateOptions{
		CheckChangeNeeded:       true,
		AllowMissingChangeFiles: true,
	})
	require.NoError(t, err)
	assert.True(t, result.IsChangeNeeded)
	assert.NotContains(t, buf.String(), "ERROR:")
}

File: rust/tests/validate_test.rs
mod common;

use beachball::options::get_options::get_parsed_options_for_test;
use beachball::types::options::{BeachballOptions, CliOptions};
use beachball::validation::validate::{ValidateOptions, ValidationError, validate};
use common::DEFAULT_REMOTE_BRANCH;
use common::repository::Repository;
use common::repository_factory::RepositoryFactory;
use common::{capture_logging, get_log_output, reset_logging};

fn validate_wrapper(
    repo: &Repository,
    validate_options: ValidateOptions,
) -> Result<beachball::validation::validate::ValidationResult, anyhow::Error> {
    let repo_opts = BeachballOptions {
        branch: DEFAULT_REMOTE_BRANCH.to_string(),
        fetch: false,
        ..Default::default()
    };
    let parsed = get_parsed_options_for_test(repo.root_path(), CliOptions::default(), repo_opts);
    validate(&parsed, &validate_options)
}

// TS: "succeeds with no changes"
#[test]
fn succeeds_with_no_changes() {
    let factory = RepositoryFactory::new("monorepo");
    let repo = factory.clone_repository();
    repo.checkout(&["-b", "test"]);

    capture_logging();
    let result = validate_wrapper(
        &repo,
        ValidateOptions {
            check_change_needed: true,
            ..Default::default()
        },
    );
    let output = get_log_output();
    reset_logging();

    assert!(result.is_ok());
    assert!(!result.unwrap().is_change_needed);
    assert!(output.contains("Validating options and change files..."));
}

// TS: "exits with error by default if change files are needed"
#[test]
fn exits_with_error_if_change_files_needed() {
    let factory = RepositoryFactory::new("monorepo");
    let repo = factory.clone_repository();
    repo.checkout(&["-b", "test"]);
    repo.stage_change("packages/foo/test.js");

    capture_logging();
    let result = validate_wrapper(
        &repo,
        ValidateOptions {
            check_change_needed: true,
            ..Default::default()
        },
    );
    let output = get_log_output();
    reset_logging();

    let err = result.expect_err("expected validation to fail");
    assert!(err.downcast_ref::<ValidationError>().is_some());
    assert!(output.contains("ERROR: Change files are needed!"));
    assert!(output.contains("Found changes in the following packages"));
}

// TS: "returns and does not log an error if change files are needed and allowMissingChangeFiles is true"
#[test]
fn returns_without_error_if_allow_missing_change_files() {
    let factory = RepositoryFactory::new("monorepo");
    let repo = factory.clone_repository();
    repo.checkout(&["-b", "test"]);
    repo.stage_change("packages/foo/test.js");

    capture_logging();
    let result = validate_wrapper(
        &repo,
        ValidateOptions {
            check_change_needed: true,
            allow_missing_change_files: true,
            ..Default::default()
        },
    );
    let output = get_log_output();
    reset_logging();

    assert!(result.is_ok());
    assert!(result.unwrap().is_change_needed);
    assert!(!output.contains("ERROR:"));
}
